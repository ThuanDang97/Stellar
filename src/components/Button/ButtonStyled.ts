import styled, { css } from 'styled-components'

// Props type
import { ButtonProps } from '@self-types/components/Button.props'

// Utils
import { pxToRem } from '@utils/theme'

// Constants
import { SIZE_BUTTON, VARIANT_BUTTON } from '@constants/index'

export const ButtonStyled = styled.button<ButtonProps>`
  ${({ btnWidth, theme }) =>
    btnWidth &&
    `width: ${
      typeof btnWidth === 'string'
        ? btnWidth
        : pxToRem(btnWidth || theme.metrics.icons.sm)
    }`};
  ${({ btnHeight, theme }) =>
    btnHeight && `height: ${pxToRem(btnHeight || theme.metrics.icons.sm)}`};
  ${(props) => props.mTop && `margin-top: ${pxToRem(props.mTop)}`};
  ${(props) => props.mLeft && `margin-left: ${pxToRem(props.mLeft)}`};
  ${(props) => props.textAlign && `text-align: ${props.textAlign}`};
  ${(props) => props.mBottom && `margin-bottom: ${pxToRem(props.mBottom)}`};
  ${(props) => props.borderColor && `border-color: ${props.borderColor};`}
  ${(props) =>
    props.letterSpacing && `letter-spacing: ${pxToRem(props.letterSpacing)};`}
    cursor: pointer;
  &:disabled {
    cursor: no-drop;
  }
  &:hover {
    ${(props) =>
      props.bgColorHover && `background-color: ${props.bgColorHover};`}
  }

  ${(props) => {
    switch (props.variant) {
      case VARIANT_BUTTON.PRIMARY:
        return css`
          color: ${({ theme }) => theme.colors.tangaroa};
          background-color: ${({ theme }) => theme.colors.white};
          border-width: ${({ theme }) =>
            pxToRem(theme.metrics.dimensions.tiny)};
          border-radius: ${({ theme }) =>
            pxToRem(theme.metrics.borderRadius.default)};
          width: ${pxToRem(85)};
          border: ${({ theme }) => pxToRem(theme.metrics.dimensions.tiny)} solid
            ${({ theme }) => theme.colors.black};
          @media (max-width: ${({ theme }) =>
              pxToRem(theme.metrics.breakPoints.xs)}) {
            width: ${({ theme }) => pxToRem(theme.metrics.width.md)};
            padding: ${({ theme }) => pxToRem(theme.metrics.dimensions.sm)} 0;
          }
        `
      case VARIANT_BUTTON.NO_BORDER:
        return css`
          color: ${({ theme }) => theme.colors.cello};
          background-color: ${({ theme }) => theme.colors.white};
          border: none;
        `
      case VARIANT_BUTTON.DARK:
        return css`
          color: ${({ theme }) =>
            props.color ? props.color : theme.colors.cello};
          background: none;
          border: none;
        `
      case VARIANT_BUTTON.DANGER:
        return css`
          color: ${({ theme }) => theme.colors.pomegranate};
          background-color: transparent;
          border: none;
        `
      case VARIANT_BUTTON.LIGHT:
        return css`
          color: ${({ theme }) => theme.colors.white};
          background: none;
          border: none;
        `
      case VARIANT_BUTTON.DISABLED:
        return css`
          color: ${({ theme }) =>
            props.color ? props.color : theme.colors.white};
          background-color: ${({ theme }) =>
            props.bgColor ? props.bgColor : theme.colors.silver};
          border: none;
          border-radius: ${({ theme }) => pxToRem(theme.metrics.dimensions.xs)};
        `
      case VARIANT_BUTTON.ICON:
        return css`
          position: relative;
          background-color: transparent;
          border: none;

          @media (max-width: ${({ theme }) =>
              pxToRem(theme.metrics.breakPoints.xs)}) {
            margin-left: ${props.mLeftMobile
              ? pxToRem(props.mLeftMobile)
              : props.theme.metrics.dimensions.tiny};
          }
        `
      case VARIANT_BUTTON.WARNING:
        return css`
          border: none;
          color: ${({ theme }) =>
            props.color ? props.color : theme.colors.white};
          background-color: ${({ theme }) =>
            props.bgColor ? props.bgColor : theme.colors.pomegranate};
          border-radius: ${({ theme }) => pxToRem(theme.metrics.dimensions.xs)};
        `
      case VARIANT_BUTTON.HALF_CIRCLE:
        return css`
          border: none;
          color: ${({ theme }) =>
            props.color ? props.color : theme.colors.white};
          background-color: ${({ theme }) =>
            props.bgColor ? props.bgColor : theme.colors.cello};
          border-radius: ${({ theme }) => pxToRem(theme.metrics.dimensions.lg)};
        `
      default:
        return css`
          color: ${({ theme }) => theme.colors.mercury};
          background-color: ${({ theme }) => theme.colors.shark};
          border-radius: ${({ theme }) => pxToRem(theme.metrics.dimensions.xs)};
          border: none;
        `
    }
  }}

  ${(props) => {
    switch (props.size) {
      case SIZE_BUTTON.NO_PADDING:
        return css`
          font-size: ${({ theme }) =>
            pxToRem(theme.typography.fontSize.common)};
          padding: 0;
        `
      case SIZE_BUTTON.MEDIUM:
        return css`
          font-size: ${({ theme }) =>
            pxToRem(theme.typography.fontSize.common)};
          padding: ${({ theme }) => pxToRem(theme.metrics.dimensions.sm)}
            ${({ theme }) => pxToRem(theme.metrics.dimensions.md)};
          @media (max-width: ${({ theme }) =>
              pxToRem(theme.metrics.breakPoints.xs)}) {
            font-size: ${({ theme }) => pxToRem(theme.typography.fontSize.sm)};
          }
        `
      case SIZE_BUTTON.LARGE:
        return css`
          font-size: ${({ theme }) =>
            pxToRem(theme.typography.fontSize.common)};
          padding: ${({ theme }) => pxToRem(theme.metrics.dimensions.base)}
            ${pxToRem(50)};
          @media (max-width: ${({ theme }) =>
              pxToRem(theme.metrics.breakPoints.xs)}) {
            font-size: ${({ theme }) => pxToRem(theme.typography.fontSize.sm)};
          }
        `
      default:
        return css`
          font-size: ${({ theme }) =>
            props.fontSize
              ? pxToRem(props.fontSize)
              : pxToRem(theme.typography.fontSize.sm)};
          font-weight: ${({ theme }) =>
            props.fontWeight
              ? props.fontWeight
              : theme.typography.fontWeight.base};
          padding: ${({ theme }) =>
            props.padding
              ? `${pxToRem(theme.metrics.dimensions.sm)} ${pxToRem(
                  props.padding,
                )}`
              : pxToRem(theme.metrics.dimensions.sm)};
          @media (max-width: ${({ theme }) =>
              pxToRem(theme.metrics.breakPoints.xs)}) {
            font-size: ${({ theme }) =>
              pxToRem(theme.typography.fontSize.base)};
          }
        `
    }
  }}
`

export const ButtonIconStyled = styled.span`
  padding-right: ${({ theme }) => pxToRem(theme.metrics.dimensions.tiny)};
`
