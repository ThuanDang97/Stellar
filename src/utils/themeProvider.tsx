import { ThemeProvider } from 'styled-components'
import { render } from '@testing-library/react'
import { theme } from '@themes/index'

export const renderWithTheme = (component: React.ReactNode) => {
  return render(<ThemeProvider theme={theme}>{component}</ThemeProvider>)
}
