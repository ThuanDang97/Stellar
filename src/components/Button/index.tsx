// Library
import React, { memo } from 'react'

// Styles
import { ButtonStyled, ButtonIconStyled } from '@components/Button/ButtonStyled'

// Props type
import { ButtonProps } from '@self-types/components/Button.props'

// Themes
import { theme } from '@themes/index'

// Components
import Image from 'next/image'
import { BadgeNotifStyled, ViewStyled } from '@components/styled-components'

const Button = ({
  onClick,
  title,
  label,
  size = 'default',
  variant,
  imgUrl,
  disabled,
  width,
  height,
  count = 0,
  titleHover,
  ...props
}: ButtonProps) => (
  <ButtonStyled
    data-testid="button"
    onClick={onClick}
    size={size}
    variant={disabled ? 'disabled' : variant}
    disabled={disabled}
    title={titleHover}
    {...props}
  >
    {!!imgUrl && (
      <ButtonIconStyled>
        <ViewStyled
          position="relative"
          display="inline-block"
          width={width}
          height={height}
          mRight={title || label ? theme.metrics.dimensions.xs : 0}
        >
          <Image layout="fill" src={imgUrl} />

          {count > 0 && <BadgeNotifStyled>{count}</BadgeNotifStyled>}
        </ViewStyled>
        {label}
      </ButtonIconStyled>
    )}
    {title}
  </ButtonStyled>
)

// Check re-render for Button component
const buttonPropsAreEqual = (
  prevButton: ButtonProps,
  nextButton: ButtonProps,
) =>
  prevButton.title === nextButton.title &&
  prevButton.label === nextButton.label &&
  prevButton.variant === nextButton.variant &&
  prevButton.width === nextButton.width &&
  prevButton.height === nextButton.height &&
  prevButton.imgUrl === nextButton.imgUrl &&
  prevButton.disabled === nextButton.disabled &&
  prevButton.onClick === nextButton.onClick

export default memo(Button, buttonPropsAreEqual)
