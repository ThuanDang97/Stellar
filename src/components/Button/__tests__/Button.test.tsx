// Library
import { cleanup, fireEvent, screen } from '@testing-library/react'

// Utils
import { renderWithTheme } from '@utils/themeProvider'

// Components
import Button from '@components/Button'

// Define props
const props = {
  label: 'Button',
  onClick: jest.fn(),
}

describe('Button render', () => {
  afterEach(cleanup)

  const component = renderWithTheme(<Button {...props} />)

  it('should render button component', () => {
    const { asFragment } = component

    expect(asFragment()).toMatchSnapshot()
  })

  it('should call onClick when click button', async () => {
    renderWithTheme(<Button {...props} />)

    fireEvent.click(await screen.findByTestId('button'))

    expect(props.onClick).toHaveBeenCalled()
  })
})
