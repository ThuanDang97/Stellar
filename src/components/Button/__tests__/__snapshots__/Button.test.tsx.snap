// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button render should render button component 1`] = `
<DocumentFragment>
  <button
    class="ButtonStyled-sc-1ys9v52-0 eiCsWB"
    data-testid="button"
  />
</DocumentFragment>
`;
